File: CS/MultiSelectColumnCustomization/GridViewOverride.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors.Customization;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.Customization;

namespace MultiSelectColumnCustomization
{
	public class GridViewOverride : GridView
	{
		public GridViewOverride(GridControl ownerGrid)
			: base(ownerGrid)
		{
		}
		public GridViewOverride()
		{
		}

		protected override CustomizationForm CreateCustomizationForm()
		{
			return new MultiSelectCustomizationForm(this);
		}

		protected override string ViewName
		{
			get
			{
				return "GridViewOverride";
			}
		}
	}

	public class MultiSelectCustomizationForm : CustomizationForm
	{
		public MultiSelectCustomizationForm(GridView view)
			: base(view)
		{
		}

		protected override CustomCustomizationListBoxBase CreateCustomizationListBox()
		{
			return new MultiSelectColumnCustomizationListBox(this);
		}

		protected override void CreateListBox()
		{
			base.CreateListBox();

			Panel bottomPanel = new Panel();
			bottomPanel.Parent = this;
			bottomPanel.Dock = DockStyle.Bottom;
			bottomPanel.Height = 30;
			bottomPanel.SendToBack();

			Button bAddCheckedCols = new Button();
			bAddCheckedCols.Parent = bottomPanel;
			bAddCheckedCols.Dock = DockStyle.Fill;
			bAddCheckedCols.Text = "Add checked columns to grid";

			bAddCheckedCols.Click += new EventHandler(OnButtonAddCheckedColumns_Click);
		}

		private void OnButtonAddCheckedColumns_Click(object sender, EventArgs e)
		{
			MultiSelectColumnCustomizationListBox listBox = (MultiSelectColumnCustomizationListBox)ActiveListBox;
			for ( int i = 0; i < listBox.CheckedItems.Count; i++ )
			{
				if ( listBox.CheckedItems[i] != null )
					((GridColumn)listBox.CheckedItems[i]).Visible = true;
			}
		}
	}

	public class MultiSelectColumnCustomizationListBox : ColumnCustomizationListBox
	{
		private int pushedIndex = -1;
		List<object> checkedItems = new List<object>();
		object focusedItem = null;

		public MultiSelectColumnCustomizationListBox(CustomizationForm form)
			: base(form)
		{
		}

		protected Rectangle CalcCheckBoxRect(GraphicsCache cache, Rectangle bounds)
		{
			RepositoryItemCheckEdit checkBox = new RepositoryItemCheckEdit();
			CheckEditViewInfo viewInfo = (CheckEditViewInfo)checkBox.CreateViewInfo();
			viewInfo.Bounds = bounds;
			viewInfo.CalcViewInfo(cache.Graphics);
			ControlGraphicsInfoArgs args = new ControlGraphicsInfoArgs(viewInfo, cache, bounds);
			CheckObjectInfoArgs ci = ((CheckEditViewInfo)args.ViewInfo).CheckInfo;

			Rectangle retValue = new Rectangle(bounds.X, ci.GlyphRect.Y, ci.GlyphRect.Width, ci.GlyphRect.Height);
			return retValue;
		}

		protected override void DrawItemObject(GraphicsCache cache, int index, Rectangle bounds)
		{
			Rectangle checkBoxRect = CalcCheckBoxRect(cache, bounds);
			bounds.X += checkBoxRect.Width;
			bounds.Width -= checkBoxRect.Width;

			base.DrawItemObject(cache, index, bounds);

			ButtonState checkState = ButtonState.Normal;

			if ( index == pushedIndex )
				checkState = ButtonState.Pushed;

			if ( checkedItems.Contains(Items[index]) )
				checkState = ButtonState.Checked;

			cache.Paint.DrawCheckBox(cache.Graphics, checkBoxRect, checkState);
			if ( focusedItem != null && Items.IndexOf(focusedItem) == index )
				cache.Paint.DrawFocusRectangle(cache.Graphics, checkBoxRect, CustomizationForm.ForeColor, CustomizationForm.BackColor);
		}

		protected override void OnMouseDown(MouseEventArgs e)
		{
			Point mousePoint = new Point(e.X, e.Y);
			object pointedItem = ItemByPoint(new Point(e.X, e.Y));
			int itemIndex = Items.IndexOf(pointedItem);
			Rectangle itemRect = GetItemRectangle(itemIndex);
			Rectangle checkBoxRect = CalcCheckBoxRect(new GraphicsCache(this.CreateGraphics()), itemRect);

			if ( checkBoxRect.Contains(mousePoint) && e.Button == MouseButtons.Left )
			{
				pushedIndex = itemIndex;
				this.InvalidateObject(pointedItem);

				return;
			}
			else if ( !checkBoxRect.Contains(mousePoint) )
				base.OnMouseDown(e);
		}

		protected override void OnMouseUp(MouseEventArgs e)
		{
			Point mousePoint = new Point(e.X, e.Y);
			object pointedItem = ItemByPoint(new Point(e.X, e.Y));
			int itemIndex = Items.IndexOf(pointedItem);
			Rectangle itemRect = GetItemRectangle(itemIndex);
			Rectangle checkBoxRect = CalcCheckBoxRect(new GraphicsCache(this.CreateGraphics()), itemRect);

			if ( checkBoxRect.Contains(mousePoint) && e.Button == MouseButtons.Left )
			{
				if ( ModifierKeys == Keys.Shift )
				{
					int startIndex = Items.IndexOf(focusedItem);
					int endIndex = Items.IndexOf(pointedItem);
					bool check = !checkedItems.Contains(pointedItem);

					if ( endIndex >= startIndex )
						for ( int i = startIndex; i <= endIndex; i++ )
						{
							if ( check && !checkedItems.Contains(Items[i]) )
								checkedItems.Add(Items[i]);
							else if ( !check && checkedItems.Contains(Items[i]) )
								checkedItems.Remove(Items[i]);
						}
					else
						for ( int i = endIndex; i < startIndex; i++ )
						{
							if ( check && !checkedItems.Contains(Items[i]) )
								checkedItems.Add(Items[i]);
							else if ( !check && checkedItems.Contains(Items[i]) )
								checkedItems.Remove(Items[i]);
						}
				}
				else if ( ModifierKeys == Keys.None )
				{
					if ( checkedItems.Contains(pointedItem) )
						checkedItems.Remove(pointedItem);
					else
						checkedItems.Add(pointedItem);
				}

				focusedItem = pointedItem;
				pushedIndex = -1;
				this.InvalidateObject(pointedItem);

				return;
			}
			else if ( !checkBoxRect.Contains(mousePoint) )
				base.OnMouseUp(e);
		}
		
		public List<object> CheckedItems
		{
			get
			{
				return checkedItems;
			}
		}
	}
}

File: VB/MultiSelectColumnCustomization/GridViewOverride.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors.Customization
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.Customization

Namespace MultiSelectColumnCustomization
	Public Class GridViewOverride
		Inherits GridView
		Public Sub New(ByVal ownerGrid As GridControl)
			MyBase.New(ownerGrid)
		End Sub
		Public Sub New()
		End Sub

		Protected Overrides Function CreateCustomizationForm() As CustomizationForm
			Return New MultiSelectCustomizationForm(Me)
		End Function

		Protected Overrides ReadOnly Property ViewName() As String
			Get
				Return "GridViewOverride"
			End Get
		End Property
	End Class

	Public Class MultiSelectCustomizationForm
		Inherits CustomizationForm
		Public Sub New(ByVal view As GridView)
			MyBase.New(view)
		End Sub

		Protected Overrides Function CreateCustomizationListBox() As CustomCustomizationListBoxBase
			Return New MultiSelectColumnCustomizationListBox(Me)
		End Function

		Protected Overrides Sub CreateListBox()
			MyBase.CreateListBox()

			Dim bottomPanel As New Panel()
			bottomPanel.Parent = Me
			bottomPanel.Dock = DockStyle.Bottom
			bottomPanel.Height = 30
			bottomPanel.SendToBack()

			Dim bAddCheckedCols As New Button()
			bAddCheckedCols.Parent = bottomPanel
			bAddCheckedCols.Dock = DockStyle.Fill
			bAddCheckedCols.Text = "Add checked columns to grid"

			AddHandler bAddCheckedCols.Click, AddressOf OnButtonAddCheckedColumns_Click
		End Sub

		Private Sub OnButtonAddCheckedColumns_Click(ByVal sender As Object, ByVal e As EventArgs)
			Dim listBox As MultiSelectColumnCustomizationListBox = CType(ActiveListBox, MultiSelectColumnCustomizationListBox)
			For i As Integer = 0 To listBox.CheckedItems.Count - 1
				If listBox.CheckedItems(i) IsNot Nothing Then
					CType(listBox.CheckedItems(i), GridColumn).Visible = True
				End If
			Next i
		End Sub
	End Class

	Public Class MultiSelectColumnCustomizationListBox
		Inherits ColumnCustomizationListBox
		Private pushedIndex As Integer = -1
		Private checkedItems_Renamed As New List(Of Object)()
		Private focusedItem As Object = Nothing

		Public Sub New(ByVal form As CustomizationForm)
			MyBase.New(form)
		End Sub

		Protected Function CalcCheckBoxRect(ByVal cache As GraphicsCache, ByVal bounds As Rectangle) As Rectangle
			Dim checkBox As New RepositoryItemCheckEdit()
			Dim viewInfo As CheckEditViewInfo = CType(checkBox.CreateViewInfo(), CheckEditViewInfo)
			viewInfo.Bounds = bounds
			viewInfo.CalcViewInfo(cache.Graphics)
			Dim args As New ControlGraphicsInfoArgs(viewInfo, cache, bounds)
			Dim ci As CheckObjectInfoArgs = (CType(args.ViewInfo, CheckEditViewInfo)).CheckInfo

			Dim retValue As New Rectangle(bounds.X, ci.GlyphRect.Y, ci.GlyphRect.Width, ci.GlyphRect.Height)
			Return retValue
		End Function

		Protected Overrides Sub DrawItemObject(ByVal cache As GraphicsCache, ByVal index As Integer, ByVal bounds As Rectangle)
			Dim checkBoxRect As Rectangle = CalcCheckBoxRect(cache, bounds)
			bounds.X += checkBoxRect.Width
			bounds.Width -= checkBoxRect.Width

			MyBase.DrawItemObject(cache, index, bounds)

			Dim checkState As ButtonState = ButtonState.Normal

			If index = pushedIndex Then
				checkState = ButtonState.Pushed
			End If

			If checkedItems_Renamed.Contains(Items(index)) Then
				checkState = ButtonState.Checked
			End If

			cache.Paint.DrawCheckBox(cache.Graphics, checkBoxRect, checkState)
			If focusedItem IsNot Nothing AndAlso Items.IndexOf(focusedItem) = index Then
				cache.Paint.DrawFocusRectangle(cache.Graphics, checkBoxRect, CustomizationForm.ForeColor, CustomizationForm.BackColor)
			End If
		End Sub

		Protected Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
			Dim mousePoint As New Point(e.X, e.Y)
			Dim pointedItem As Object = ItemByPoint(New Point(e.X, e.Y))
			Dim itemIndex As Integer = Items.IndexOf(pointedItem)
			Dim itemRect As Rectangle = GetItemRectangle(itemIndex)
			Dim checkBoxRect As Rectangle = CalcCheckBoxRect(New GraphicsCache(Me.CreateGraphics()), itemRect)

			If checkBoxRect.Contains(mousePoint) AndAlso e.Button = MouseButtons.Left Then
				pushedIndex = itemIndex
				Me.InvalidateObject(pointedItem)

				Return
			ElseIf (Not checkBoxRect.Contains(mousePoint)) Then
				MyBase.OnMouseDown(e)
			End If
		End Sub

		Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
			Dim mousePoint As New Point(e.X, e.Y)
			Dim pointedItem As Object = ItemByPoint(New Point(e.X, e.Y))
			Dim itemIndex As Integer = Items.IndexOf(pointedItem)
			Dim itemRect As Rectangle = GetItemRectangle(itemIndex)
			Dim checkBoxRect As Rectangle = CalcCheckBoxRect(New GraphicsCache(Me.CreateGraphics()), itemRect)

			If checkBoxRect.Contains(mousePoint) AndAlso e.Button = MouseButtons.Left Then
				If ModifierKeys = Keys.Shift Then
					Dim startIndex As Integer = Items.IndexOf(focusedItem)
					Dim endIndex As Integer = Items.IndexOf(pointedItem)
					Dim check As Boolean = Not checkedItems_Renamed.Contains(pointedItem)

					If endIndex >= startIndex Then
						For i As Integer = startIndex To endIndex
							If check AndAlso (Not checkedItems_Renamed.Contains(Items(i))) Then
								checkedItems_Renamed.Add(Items(i))
							ElseIf (Not check) AndAlso checkedItems_Renamed.Contains(Items(i)) Then
								checkedItems_Renamed.Remove(Items(i))
							End If
						Next i
					Else
						For i As Integer = endIndex To startIndex - 1
							If check AndAlso (Not checkedItems_Renamed.Contains(Items(i))) Then
								checkedItems_Renamed.Add(Items(i))
							ElseIf (Not check) AndAlso checkedItems_Renamed.Contains(Items(i)) Then
								checkedItems_Renamed.Remove(Items(i))
							End If
						Next i
					End If
				ElseIf ModifierKeys = Keys.None Then
					If checkedItems_Renamed.Contains(pointedItem) Then
						checkedItems_Renamed.Remove(pointedItem)
					Else
						checkedItems_Renamed.Add(pointedItem)
					End If
				End If

				focusedItem = pointedItem
				pushedIndex = -1
				Me.InvalidateObject(pointedItem)

				Return
			ElseIf (Not checkBoxRect.Contains(mousePoint)) Then
				MyBase.OnMouseUp(e)
			End If
		End Sub

		Public ReadOnly Property CheckedItems() As List(Of Object)
			Get
				Return checkedItems_Renamed
			End Get
		End Property
	End Class
End Namespace